matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/worker/MatrixWorkerFactory.kt

```diff
@@ -20,25 +20,64 @@ import android.content.Context
+/**
+ * This factory is responsible of creating Workers by giving the session manager.
+ * This is not the cleanest way but getting SessionComponent is dependant of args type.
+ */
+@MatrixScope
+internal class MatrixWorkerFactory @Inject constructor(private val sessionManager: SessionManager) : WorkerFactory() {
 
     override fun createWorker(
             appContext: Context,
             workerClassName: String,
             workerParameters: WorkerParameters
     ): ListenableWorker? {
         Timber.d("MatrixWorkerFactory.createWorker for $workerClassName")
-
-        val foundEntry =
-                workerFactories.entries.find { Class.forName(workerClassName).isAssignableFrom(it.key) }
-        val factoryProvider = foundEntry?.value
-                ?: throw IllegalArgumentException("unknown worker class name: $workerClassName")
-        return factoryProvider.get().create(appContext, workerParameters)
+        return when (workerClassName) {
+            AddPusherWorker::class.java.name                      ->
+                AddPusherWorker(appContext, workerParameters, sessionManager)
+            CancelGossipRequestWorker::class.java.name            ->
+                CancelGossipRequestWorker(appContext, workerParameters, sessionManager)
+            GetGroupDataWorker::class.java.name                   ->
+                GetGroupDataWorker(appContext, workerParameters, sessionManager)
+            MultipleEventSendingDispatcherWorker::class.java.name ->
+                MultipleEventSendingDispatcherWorker(appContext, workerParameters, sessionManager)
+            RedactEventWorker::class.java.name                    ->
+                RedactEventWorker(appContext, workerParameters, sessionManager)
+            SendEventWorker::class.java.name                      ->
+                SendEventWorker(appContext, workerParameters, sessionManager)
+            SendGossipRequestWorker::class.java.name              ->
+                SendGossipRequestWorker(appContext, workerParameters, sessionManager)
+            SendGossipWorker::class.java.name                     ->
+                SendGossipWorker(appContext, workerParameters, sessionManager)
+            SendVerificationMessageWorker::class.java.name        ->
+                SendVerificationMessageWorker(appContext, workerParameters, sessionManager)
+            SyncWorker::class.java.name                           ->
+                SyncWorker(appContext, workerParameters, sessionManager)
+            UpdateTrustWorker::class.java.name                    ->
+                UpdateTrustWorker(appContext, workerParameters, sessionManager)
+            UploadContentWorker::class.java.name                  ->
+                UploadContentWorker(appContext, workerParameters, sessionManager)
+            else                                                  ->
+                // Return null to delegate to the default WorkerFactory.
+                null
```

Comment:
Maybe log a warning for the developers, if they forget to add their new worker instanciation above?

Reply:
Yes good point